Blob placeholders prototype

ChumpChief opened the pull request:
Prototype branch for blob placeholders.  Replaces #24137 which has more context.

ChumpChief committed:
Initial impl and plumbing

ChumpChief committed:
Handle placeholder in resolving handles

ChumpChief committed:
Random is fun

ChumpChief committed:
Enable experimental behavior by default

ChumpChief committed:
Remove explicit detached failing

ChumpChief committed:
Skip failing tests

anthony-murphy committed:
Blob Placeholder: Enable no wait (#24157)

This is just for testing purposes. We intentionally make it very hard to
observe local object creation, and attach state, in a broad way. Test
like the local server stress work around this by storing absolute paths,
which allow remote clients to find objects one they are attached but
avoid attaching them at that point. Clients walk the list of absolute
paths, and find those they can resolve, and then use those in dds
operations, this can lead to attach, multiple handles to the same
object, and reviving unreferenced objects.

ChumpChief committed:
Skip 2 more seeds

ChumpChief committed:
Merge remote-tracking branch 'upstream/main' into test/blob-placeholders

ChumpChief committed:
Drop separate experimental API in favor of feature flag, and support detached container flow (#24179)

anthony-murphy commented on packages/runtime/container-runtime/src/blobManager/blobManager.ts:
```diff
@@ -435,6 +452,15 @@ export class BlobManager {
 		if (this.runtime.attachState === AttachState.Detached) {
 			return this.createBlobDetached(blob);
 		}
+		return this.mc.config.getBoolean("Fluid.Runtime.UploadBlobPlaceholders") === true
```
what is the test strategy for this? currently its not enabled for any tests.

ChumpChief replied:
Yeah I need to figure that out still.  Haven't looked at how other feature flags do it yet, but seems like it would be a waste of time to test both on/off for most of our tests.  At first glance looks like the blob and GC tests are the main interesting ones.

anthony-murphy replied:
we will have production users on both behaviors, seem like we need strong guarantees they both continue to work until we only have one behavior.

anthony-murphy commented on packages/runtime/container-runtime/src/blobManager/blobManager.ts:
```diff
@@ -475,6 +501,27 @@ export class BlobManager {
+	private createBlobPlaceholder(blob: ArrayBufferLike): IFluidHandleInternal<ArrayBufferLike> {
+		const localId = this.localBlobIdGenerator();
+
+		return new BlobHandle(
```
how does this work while attaching?

ChumpChief replied:
Not sure I follow the question - while attaching what specifically?

anthony-murphy replied:
look at createBlobLegacy, it waits to upload while attaching

ChumpChief replied:
Sure, we can do the same in the new flow.

ChumpChief replied:
Longer-term maybe we should look into why this was needed, since it'll block making the API sync if we ever want to do that.

ChumpChief replied:
#24184

anthony-murphy commented on packages/runtime/container-runtime/src/blobManager/blobManager.ts:
```diff
@@ -374,8 +379,20 @@ export class BlobManager {
 			storageId = blobId;
 		} else {
 			const attachedStorageId = this.redirectTable.get(blobId);
-			assert(!!attachedStorageId, 0x11f /* "requesting unknown blobs" */);
```
well probably need something more robust here which can differentiate between deferred vs non-deferred handles

ChumpChief replied:
Agreed it will be needed long-term (particularly thinking about observability), but are you saying you think it's also needed for the more immediate problem we're trying to solve?  Is your main concern around keeping this assert enabled for non-deferred handles?

anthony-murphy replied:
yeah. i think we should be able to differentiate between the new vs old handles

ChumpChief replied:
OK but do you have a particular purpose in mind?  I think you are hinting that you have a goal in mind for this capability, want to make sure I'm on the same page

anthony-murphy replied:
the immediate goal is to keep the assert, or an error for blobs which we know should exists, and only suppress when there are blobs with the new behavior

ChumpChief replied:
This was resolved by #24224

dannimad commented:
Is there any design doc, one pager or context to help review this approach? I believe I get the general idea through spare conversations but a written starting point would be great and I'd appreciate a pointer in case it already exists.

ChumpChief committed:
Enable gc dedupe tests and delay returning the handle if in Attaching state (#24184)

ChumpChief committed:
Merge branch 'main' into test/blob-placeholders

ChumpChief committed:
Test both with/without blob placeholders enabled (#24212)

ChumpChief committed:
Enable client build pipeline for test/blob-placeholders branch (#24214)

ChumpChief committed:
Add metadata to handles and use in BlobManager for placeholder blobs (#24224)

ChumpChief committed:
Switch to using DocumentSchema to enable blob placeholders (#24262)

DocumentSchema is a more appropriate tool than a feature flag, because
once it is enabled for a document it permanently puts older incompatible
clients at risk of encountering op patterns they aren't prepared for
(namely, an op containing the handle before the blob attach op).

This PR primarily switches to that mechanism, but it also revealed that
I had a bug in my flag enablement in blobManager.spec.ts that was
causing the tests to run only in legacy mode (I wasn't paying attention
to how/when injectedSettings got wiped out). After fixing that test bug,
the now-functioning tests revealed a couple other small bugs:

1. The pendingBlobs created by the placeholder handles should be marked
attached immediately upon creation, because they are created by the act
of attaching the handle.
2. waitForBlob in the test had an inverted condition - it should be
waiting to see its blob appear in the unprocessedBlobs (exists in main
but must not be causing problems there)
3. An 0x38f assert in one of the tests that I haven't debugged fully
yet, though it may just be a test bug due to the handle management.

Because blobManager.spec.ts expects to have such intimate knowledge of
the BlobManager internals, it might be reasonable to consider a broader
rework of its test coverage for placeholder blobs (since it subverts the
expectations about handle and upload behavior). Maybe for later
consideration.


[AB#34418](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/34418)

[AB#34459](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/34459)

scottn12 committed:
Revert "Switch to using DocumentSchema to enable blob placeholders (#… (#24266)

…24262)"

This reverts commit 68e75aecbab81fed453326afcc97e140503fcf1d.

ChumpChief committed:
Switch to using DocumentSchema to enable blob placeholders (#24268)

ChumpChief committed:
Merge remote-tracking branch 'upstream/main' into test/blob-placeholders

ChumpChief committed:
Replace metadata with placeholder tagging (#24308)

ChumpChief committed:
Merge remote-tracking branch 'upstream/main' into test/blob-placeholders

ChumpChief committed:
Add observability API surface for locally created BlobHandles (#24397)

Primary motivation for this change is to ensure customers can still
detect upload failures (so there's no feature take-back as compared to
the non-placeholder case).

Main changes are:
* Adds the interfaces to core-interfaces as legacy/alpha
* Adds a type check to runtime-utils as legacy/alpha
* Update BlobManager/BlobHandle to implement
* Add/update two tests for the local/failed/shared states and their
transitions (placeholder state is unused until we do remote handles too)
* Small rename to be more normally readable
(IFluidHandleInternalPlaceholder -> IFluidPlaceholderHandleInternal)


[AB#34918](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/34918)
